### PR TITLE
[bugfix] Add missing location information in errors for functions inside element constructors (err:XQTY0105)

### PR DESCRIPTION
**Pull request by chris-evolvedbinary**
### Description:

Add location information to map expressions, and pass this information to the `XQTY0105` exception that occurs when an enclosed expression contains a function item

### Reference:

https://github.com/eXist-db/exist/issues/3623

### Type of tests:

`subexpression-in-enclosed-expression-evaluates-to-map` and `enclosed-expression-evaluates-to-map` in `exist-core\src\test\xquery\errors.xql`

--

This open source contribution to the eXist-db project was commissioned by the Office of the Historian, U.S. Department of State, https://history.state.gov/.

### CODE REVIEW

File: exist-core/src/main/java/org/exist/xquery/value/FunctionReference.java [hunk 1]
```diff
@@ -170,4 +170,25 @@ public AtomicValue min(Collator collator, AtomicValue other)
     public AtomicValue atomize() throws XPathException {
         throw new XPathException(ErrorCodes.FOTY0013, "A function item other than an array cannot be atomized");
     }
+
+    /** the expression from which this value derives */
+    private Expression expression = null;
```
Comment (dizzzz) [inline exist-core/src/main/java/org/exist/xquery/value/FunctionReference.java:175]: is this null assignment really needed?

**Review by dizzzz — APPROVED** [review]

**Comment by dizzzz** [comment]
image

**Review by adamretter — CHANGES_REQUESTED** [review]
This solves the problem for a Map expr inside an element constructor. But I think there are several more types of expr and function that we need to at least add tests for, and then fix potentially if they don't work already:

1. What about `ArrayExpr`?
2. What about evaluation of a`FunctionReference`? e.g. `let $f := function($x) { fn:error(xs:QName("local:my-error") } return element { $x() }`
3. What about `UserDefinedFunction`? i.e. functions implemented in XQuery? You could implement an XQuery function that wraps `fn:error` for testing this.
4. What about `BasicFunction`? i.e. functions for XQuery implemented in Java? You could use fn:error for testing this
5. What about other types of expression inside the element constructor... as just one example: Attribute Constructor:
`element e1 { attribute a1 { fn:error(xs:QName("local:my-error")) } }`

**Comment by chris-evolvedbinary** [comment]
> This solves the problem for a Map expr inside an element constructor. But I think there are several more types of expr and function that we need to at least add tests for, and then fix potentially if they don't work already:
> 
> 1. What about `ArrayExpr`?
> 2. What about evaluation of a`FunctionReference`? e.g. `let $f := function($x) { fn:error(xs:QName("local:my-error") } return element { $x() }`
> 3. What about `UserDefinedFunction`? i.e. functions implemented in XQuery? You could implement an XQuery function that wraps `fn:error` for testing this.
> 4. What about `BasicFunction`? i.e. functions for XQuery implemented in Java? You could use fn:error for testing this
> 5. What about other types of expression inside the element constructor... as just one example: Attribute Constructor:
>    `element e1 { attribute a1 { fn:error(xs:QName("local:my-error")) } }`

The `exist-core/src/test/xquery/errors.xql` file already contains tests for these cases, except for the last case ("other types of expression inside the element constructor") for which I added a test.  All the tests pass.

The issue is that the `UserDefinedFunction`, `BasicFunction`, and `OpNumeric` class all derive from the `PathExpr` class, which has a list of expressions (as "steps").  So when errors occur for a `PathExpr` object, the location information is obtainable for an expression of the object.  The `MapType` class, however, derives from the `FunctionReference` class, which does not derive from the `PathExpr` class and therefore does not have information about the expression from which it came.  So the code for this pull request adds an expression property to the `FunctionReference` class.

Also, there does not appear to be an `ArrayExpr` class.  Instead, the `XQueryTreeParser` class uses the `ArrayConstructor` class, which like the `PathExpr` class has a list of expressions (as "arguments").

**Comment by adamretter** [comment]
@chris-evolvedbinary I have pushed one commit which adds some additional tests to cover the issues I mentioned. Hopefully that is helpful?

4 of the tests I added do fail, so I think there are one or two more places where you need to expose the location information.

**Review by adamretter — APPROVED** [review]
LGTM

**Pull request merged**